src/drivers/dshot/dshot.cpp

```diff
@@ -0,0 +1,1679 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include <float.h>
+#include <math.h>
+
+#include <board_config.h>
+#include <drivers/device/device.h>
+#include <drivers/drv_hrt.h>
+#include <drivers/drv_input_capture.h>
+#include <drivers/drv_mixer.h>
+#include <drivers/drv_pwm_output.h>
+#include <lib/cdev/CDev.hpp>
+#include <lib/mathlib/mathlib.h>
+#include <lib/mixer_module/mixer_module.hpp>
+#include <lib/parameters/param.h>
+#include <lib/perf/perf_counter.h>
+#include <px4_arch/dshot.h>
+#include <px4_atomic.h>
+#include <px4_config.h>
+#include <px4_getopt.h>
+#include <px4_log.h>
+#include <px4_module.h>
+#include <px4_platform_common/px4_work_queue/ScheduledWorkItem.hpp>
+#include <uORB/Publication.hpp>
+#include <uORB/PublicationMulti.hpp>
+#include <uORB/Subscription.hpp>
+#include <uORB/SubscriptionCallback.hpp>
+#include <uORB/topics/actuator_armed.h>
+#include <uORB/topics/actuator_controls.h>
+#include <uORB/topics/actuator_outputs.h>
+#include <uORB/topics/multirotor_motor_limits.h>
+#include <uORB/topics/parameter_update.h>
+#include <uORB/topics/esc_status.h>
+
+#include "telemetry.h"
+
+
+using namespace time_literals;
+
+/** Mode given via CLI */
+enum PortMode {
+	PORT_MODE_UNSET = 0,
+	PORT_FULL_GPIO,
+	PORT_FULL_PWM,
+	PORT_PWM8,
+	PORT_PWM6,
+	PORT_PWM5,
+	PORT_PWM4,
+	PORT_PWM3,
+	PORT_PWM2,
+	PORT_PWM1,
+	PORT_PWM3CAP1,
+	PORT_PWM4CAP1,
+	PORT_PWM4CAP2,
+	PORT_PWM5CAP1,
+	PORT_PWM2CAP2,
+	PORT_CAPTURE,
+};
+
+#if !defined(BOARD_HAS_PWM)
+#  error "board_config.h needs to define BOARD_HAS_PWM"
+#endif
+
+
+class DShotOutput : public cdev::CDev, public ModuleBase<DShotOutput>, public OutputModuleInterface
+{
+public:
+	enum Mode {
```

Comment:
Same here. And is this needed?

Reply:
These are the things that come from `fmu`. We can possibly simplify, but preferably as a separate step.

Reply:
Ok, maybe we can extract it from fmu and dshot later and improve it.